// lastfm-typed/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const tag_1 = __importDefault(require("./classes/tag"));
const chart_1 = __importDefault(require("./classes/chart"));
const auth_1 = __importDefault(require("./classes/auth"));
const album_1 = __importDefault(require("./classes/album"));
const artist_1 = __importDefault(require("./classes/artist"));
const library_1 = __importDefault(require("./classes/library"));
const track_1 = __importDefault(require("./classes/track"));
const user_1 = __importDefault(require("./classes/user"));
const helper_1 = __importDefault(require("./classes/helper"));
const logger_1 = __importDefault(require("./logger"));
const events_1 = require("events");
const geo_1 = __importDefault(require("./classes/geo"));
class LastFM extends events_1.EventEmitter {
    constructor(apiKey, options) {
        var _a, _b, _c;
        super();
        if (!options) {
            options = {};
        }
        (_a = options.apiSecret) !== null && _a !== void 0 ? _a : (options.apiSecret = "");
        (_b = options.userAgent) !== null && _b !== void 0 ? _b : (options.userAgent = "lastfm-typed-npm");
        (_c = options.secureConnection) !== null && _c !== void 0 ? _c : (options.secureConnection = true);
        let { apiSecret, userAgent, secureConnection } = options;
        this.info = {
            key: apiKey,
            secret: apiSecret,
            context: this
        };
        this.tag = new tag_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.chart = new chart_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.geo = new geo_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.auth = new auth_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.album = new album_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.artist = new artist_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.library = new library_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.track = new track_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.user = new user_1.default(apiKey, this, apiSecret, userAgent, secureConnection);
        this.helper = new helper_1.default(this);
        this.logger = new logger_1.default(this);
    }
}
exports.default = LastFM;
